### PR TITLE
Enable use of global secp256k1 context

### PR DESCRIPTION
**Pull request by davxy**
Context creation is expensive.

Use the preallocated context that ships with the library.

Follows from the work done for this https://github.com/paritytech/substrate/pull/10798

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 These are the results when building the `examples/*` contracts from this branch with `cargo-contract 0.17.0-b2a35cd` and comparing them to ink! `master`: 

|                      | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:----------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`        |                  |            | 1.44 K               |                |
| `adder`              |                  |            | 2.94 K               |                |
| `contract-terminate` |                  |            | 1.31 K               | 200_778        |
| `contract-transfer`  |                  |            | 7.98 K               | 1_399          |
| `delegator`          |                  |            | 7.63 K               | 10_296         |
| `dns`                |                  |            | 10.20 K              | 15_153         |
| `erc1155`            |                  |            | 27.50 K              | 47_811         |
| `erc20`              |                  |            | 9.80 K               | 10_014         |
| `erc721`             |                  |            | 13.70 K              | 34_702         |
| `flipper`            |                  |            | 1.76 K               | 347            |
| `incrementer`        |                  |            | 1.64 K               | 340            |
| `multisig`           |                  |            | 26.79 K              | 56_756         |
| `proxy`              |                  |            | 3.81 K               | 2_629          |
| `rand-extension`     |                  |            | 5.12 K               | 6_654          |
| `subber`             |                  |            | 2.96 K               |                |
| `trait-erc20`        |                  |            | 10.10 K              | 10_659         |
| `trait-flipper`      |                  |            | 1.53 K               | 342            |
| `trait-incrementer`  |                  |            | 1.62 K               | 680            |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/178130) | Last update: Mon Feb  7 16:53:51 CET 2022

**Comment by xgreenx** [comment]
ink! doesn't use `secp256k1` to recover, generate keys. So it doesn't use the context. Usage of the context will increase the size of the contracts. It is why that function was moved to `contract-pallet`.

Maybe I'm not right and it doesn't affect the size of the contract, could you check it, please?=)

**Comment by davxy** [comment]
> ink! doesn't use `secp256k1` to recover, generate keys. So it doesn't use the context. Usage of the context will increase the size of the contracts. It is why that function was moved to `contract-pallet`.
> 
> Maybe I'm not right and it doesn't affect the size of the contract, could you check it, please?=)

@xgreenx Ink! has already replaced `libsecp256k1` with `secp256k1` here https://github.com/paritytech/ink/pull/1052

If you check the changed files, this PR only introduces the usage of the global `secp256k1` feature

**Comment by xgreenx** [comment]
Ohh, okay, you only enabled that feature for tests. You didn't touch [the code related to contracts](https://github.com/paritytech/ink/blob/master/crates/eth_compatibility/Cargo.toml#L24).

File: crates/env/Cargo.toml [hunk 1]
```diff
@@ -40,7 +40,7 @@ sha3 = { version = "0.10", optional = true }
 blake2 = { version = "0.10", optional = true }
 
 # ECDSA for the off-chain environment.
-secp256k1 = { version = "0.21.2", features = ["recovery"] }
+secp256k1 = { version = "0.21.2", features = ["recovery", "global-context"] }
```
Comment (HCastano) [inline crates/env/Cargo.toml:43]: So from what I can tell `global-context` implies `std`, so we should probably have this as `optional = true`

File: crates/engine/Cargo.toml [hunk 2]
```diff
@@ -23,7 +23,7 @@ sha3 = { version = "0.10" }
 blake2 = { version = "0.10" }
 
 # ECDSA for the off-chain environment.
-secp256k1 = { version = "0.21.2", features = ["recovery"] }
+secp256k1 = { version = "0.21.2", features = ["recovery", "global-context"] }
```
Comment (HCastano) [inline crates/engine/Cargo.toml:26]: Same comment here

**Review by HCastano — COMMENTED** [review]
Looks fine, just one thing to check related to `std` features

**Review by HCastano — APPROVED** [review]
Not sure why there are contract size changes. I checked PR out and built the contracts locally and they're smaller than what the waterfall reports (and match `master`) 🤷‍♂️ 

Anyways, thanks!

**Pull request merged**